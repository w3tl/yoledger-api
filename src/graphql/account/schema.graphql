extend type Query {
  accounts(type: AccountType!): [Account]
  account(id: ID!): Account
}

extend type Mutation {
  addAccount(input: AddAccountInput!): AddAccountPayload
}

type Account {
  id: ID!
  name: String!
  balance: Float
  type: AccountType!
}

enum AccountType {
  ASSET
  EXPENSE
  INCOME
  VENDOR
}

input AddAccountInput {
  name: String!
  balance: Float
  type: AccountType!
}

type AddAccountPayload {
  account: Account
}
