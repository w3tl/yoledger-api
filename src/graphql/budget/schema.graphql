extend type Query {
  budgets(dateStart: Date!, dateEnd: Date!): Budgets
  budget(account: String!, dateStart: Date!, dateEnd: Date!): [Budget]
}
extend type Mutation {
  upsertBudget(input: UpsertBudgetInput!): UpsertBudgetPayload
}

type Budgets {
  accounts: [Account]
}

type Budget {
  id: ID!
  date: Date!
  account: Account!
  amount: Float
}

input UpsertBudgetInput {
  account: String!
  date: Date!
  amount: Float!
}

type UpsertBudgetPayload {
  success: Boolean
  budget: Budget
}
