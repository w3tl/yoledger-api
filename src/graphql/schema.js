import { Schema as userTypeDefs } from './user';
import { Schema as transactionTypeDefs } from './transaction';
import { Schema as accountTypeDefs } from './account';
import { Schema as budgetTypeDefs } from './budget';

export default `
type Query {
  dummy: Boolean
}
type Mutation {
  dummy: Boolean
}
scalar Date
${userTypeDefs}
${accountTypeDefs}
${transactionTypeDefs}
${budgetTypeDefs}
`;
