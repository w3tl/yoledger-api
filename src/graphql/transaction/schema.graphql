extend type Query {
  transactions(dateStart: Date!, dateEnd: Date, page: Int = 0, itemsPerPage: Int = 20): [Transaction]
  transaction(id: ID!): Transaction
}
extend type Mutation {
  addTransaction(input: AddTransactionInput!): AddTransactionPayload
  deleteTransaction(id: ID!): DeleteTransactionPayload
  updateTransaction(id: ID!, input: UpdateTransactionInput!): UpdateTransactionPayload
}

type Transaction {
  id: ID!
  amount: Float!
  source: Account!
  destination: Account!
  date: Date!
}
input AddTransactionInput {
  source: String!
  destination: String!
  amount: Float!
  date: Date
}
type AddTransactionPayload {
  transaction: Transaction
}
type DeleteTransactionPayload {
  success: Boolean!
  id: ID
}
input UpdateTransactionInput {
  amount: Float
  source: String
  destination: String
  date: Date
}
type UpdateTransactionPayload {
  transaction: Transaction
}
