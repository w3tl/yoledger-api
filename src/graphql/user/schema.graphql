extend type Query {
  user: User
}

extend type Mutation {
  signin(login: String!, password: String!): SignupPayload!
  signout(token: String): SignoutPayload!
  changePassword(oldPassword: String!, newPassword: String!): ChangePasswordPayload!
}

type User {
  username: String
}

type SignupPayload {
  token: String!
}

type SignoutPayload {
  success: Boolean
}

type ChangePasswordPayload {
  token: String
}
