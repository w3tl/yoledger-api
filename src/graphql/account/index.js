import Schema from './schema.graphql';
import { Query, Mutation, Account } from './resolver';

export {
  Schema, Query, Mutation, Account,
};
