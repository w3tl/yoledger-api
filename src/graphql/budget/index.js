import Schema from './schema.graphql';
import {
  Query, Mutation, Budget,
} from './resolver';

export {
  Schema, Query, Mutation, Budget,
};
