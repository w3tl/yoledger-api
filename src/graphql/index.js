import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './schema';
import resolvers from './resolvers';
import dataloaders from './dataloaders';
import {
  User,
  Account,
  Transaction,
  Budget,
} from '../model';

export default makeExecutableSchema({
  typeDefs,
  resolvers,
});

export const models = (connection, userId) => ({
  userModel: new User(connection.db(), userId),
  accountModel: new Account(connection.db(), userId),
  transactionModel: new Transaction(connection.db(), userId),
  budgetModel: new Budget(connection.db(), userId),
});

export { dataloaders };
