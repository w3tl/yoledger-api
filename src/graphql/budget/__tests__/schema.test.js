jest.mock('mongodb');

import { graphql } from 'graphql';
import schema, { dataloaders } from '../../index';
import { Budget } from '../../../model';

describe('budget schema', () => {
  const { connection } = require('mongodb');
  const budgetModel = new Budget(connection.db(), 'admin');
  const context = {
    dataloaders: dataloaders(connection),
    models: { budgetModel },
    userId: 'admin',
  };

  it('must return plan with periods and accounts', async () => {
    const query = `
    query budgets($dateStart: Date!, $dateEnd: Date!) {
      budgets(dateStart: $dateStart, dateEnd: $dateEnd) {
        accounts {
          id
          name
        }
      }
    }
    `;
    const rootValue = {};
    const variables = {
      dateStart: '2018-05-10',
      dateEnd: '2018-06-25',
    };

    const result = await graphql(schema, query, rootValue, context, variables);
    expect(result.errors).toBeUndefined();
    expect(result.data).toMatchSnapshot();
  });

  it('query budget must return budget for account', async () => {
    const query = `
    query budget($account: String!, $dateStart: Date!, $dateEnd: Date!) {
      budget(account: $account, dateStart: $dateStart, dateEnd: $dateEnd) {
        id
        date
        account {
          id
          name
        }
        amount
      }
    }
    `;
    const rootValue = {};
    const variables = {
      account: 'Food',
      dateStart: '2018-05-10',
      dateEnd: '2018-06-25',
    };

    const result = await graphql(schema, query, rootValue, context, variables);
    expect(result.errors).toBeUndefined();
    expect(result.data).toMatchSnapshot();
  });
});
