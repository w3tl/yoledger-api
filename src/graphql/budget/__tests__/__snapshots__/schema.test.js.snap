// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`budget schema must return plan with periods and accounts 1`] = `
Object {
  "budgets": Object {
    "accounts": Array [
      Object {
        "id": "3",
        "name": "Food",
      },
      Object {
        "id": "4",
        "name": "Train",
      },
    ],
  },
}
`;

exports[`budget schema query budget must return budget for account 1`] = `
Object {
  "budget": Array [
    Object {
      "account": Object {
        "id": "3",
        "name": "Food",
      },
      "amount": 50,
      "date": "2018-05-21T00:00:00.000Z",
      "id": "1",
    },
    Object {
      "account": Object {
        "id": "3",
        "name": "Food",
      },
      "amount": 55,
      "date": "2018-05-10T00:00:00.000Z",
      "id": "2",
    },
    Object {
      "account": Object {
        "id": "3",
        "name": "Food",
      },
      "amount": 40,
      "date": "2018-06-21T00:00:00.000Z",
      "id": "3",
    },
    Object {
      "account": Object {
        "id": "3",
        "name": "Food",
      },
      "amount": 60,
      "date": "2018-06-10T00:00:00.000Z",
      "id": "4",
    },
  ],
}
`;
